src/util/string.h

```diff
@@ -17,6 +17,28 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <int num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
+    consteval static void CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
+        bool begin_f{false};
+        for (char c : str) {
+            bool perc{c == '%'};
+            if (begin_f) {
+                count += !perc;
+                begin_f = false;
+                continue;
+            }
+            if (perc) begin_f = true;
+        }
+        if (num_params != count) throw "Number of format specifiers and arguments must match";
```

Comment:
if we throw a `runtime_error`, testing becomes slightly simpler:
```C++
std::runtime_error("Format specifier count must match the argument count!")
```
otherwise it's:
```C++
BOOST_CHECK_THROW(ConstevalFormatString<1>::CheckNumFormatSpecifiers(""), const char*);
```

nit: slight rewording

Reply:
How come? Wouldn't that be a simpler way to check that we're throwing in this case (instead of the constant err and err_types which cannot even change)
```C++
BOOST_CHECK_THROW(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), const char*);
```

Reply:
Seems a bit excessive to me, could we maybe specialize the tests when those errors are actually added?